### DISCUSSION TITLE
What are some of the most used Vim shortcuts you use?

### DISCUSSION THREAD

**Discussion by jacobsayono** [community/community#78330] [Programming Help]
Hi! Just wondering how you guys use Vim, and if there are shortcuts you use most commonly at the top of your head for navigating through code, editing chunks of code, etc.

I would like to compile the most used key bindings by Vim users and make an interactive tutorial to speed up many of our workflows, especially since Vim has a high learning curve.

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @jacobsayono ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Comment by dzenebieri** [comment]
Navigation:

h, j, k, l: Move the cursor left, down, up, and right, respectively.
w, e, b: Move the cursor forward by word (w), to the end of a word (e), and backward by word (b).
0 (zero): Move to the beginning of the line.
^: Move to the first non-whitespace character in the line.
$: Move to the end of the line.
gg: Move to the beginning of the document.
G: Move to the end of the document.
Ctrl + u, Ctrl + d: Scroll half a page up and down.
Ctrl + b, Ctrl + f: Scroll a full page up and down.


Editing:

i, I: Enter insert mode before the cursor (i) or at the beginning of the line (I).
a, A: Enter insert mode after the cursor (a) or at the end of the line (A).
o, O: Open a new line below (o) or above (O) the current line in insert mode.
x: Delete the character under the cursor.
dd: Delete the current line.
yy: Yank (copy) the current line.
p, P: Paste the yanked or deleted text after (p) or before (P) the cursor.


Visual Mode:

v: Start character-wise visual mode.
V: Start line-wise visual mode.
Ctrl + v: Start block-wise visual mode.


Search and Replace:

/: Enter search mode. After typing your search term, press Enter.
n: Move to the next occurrence of the search term.
N: Move to the previous occurrence of the search term.
:%s/old/new/g: Replace all occurrences of "old" with "new" in the entire file.


Exiting Vim:

:w: Save changes.
:q: Quit (close) Vim.
:wq or ZZ: Save changes and quit.
:q!: Quit without saving changes.

Miscellaneous:

u: Undo.
Ctrl + r: Redo.
. (dot): Repeat the last change.
:help [topic]: Open help for a specific topic.
:q!: Quit without saving changes.

**Comment by madhavmadupu** [comment]
Vim shortcuts are a powerful way to navigate and edit text efficiently. Here are some commonly used ones:

1.  **Navigation:**
    
    -   **h, j, k, l:** Left, down, up, right respectively.
    -   **w, e, b:** Move forward by word, to the end of a word, to the beginning of a word respectively.
    -   **0, $:** Move to the beginning or end of a line.
    -   **gg, G:** Go to the beginning or end of the file.
    -   **Ctrl + u, Ctrl + d:** Move up or down half a page.
    -   **Ctrl + f, Ctrl + b:** Move forward or backward one page.
    -   \*_%, _:__ Jump to matching braces or symbols.
2.  **Editing:**
    
    -   **i, I:** Enter insert mode at the cursor, at the beginning of the line.
    -   **a, A:** Enter insert mode after the cursor, at the end of the line.
    -   **o, O:** Open a new line below or above the current line.
    -   **x, dd:** Delete a character, delete a line.
    -   **yy, p:** Copy a line, paste after/below the cursor.
3.  **Search and Replace:**
    
    -   **/:** Search forward.
    -   **? :** Search backward.
    -   **n, N:** Go to the next or previous search result.
    -   **:%s/old/new/g:** Replace all occurrences of 'old' with 'new' in the entire file.
4.  **Modes:**
    
    -   **Esc:** Switch to normal mode from any other mode.
    -   **v, V:** Enter visual mode (character-based, line-based).
    -   **Ctrl + v:** Enter visual block mode for column-based selection.
5.  **Tabs, Buffers, and Windows:**
    
    -   **:tabnew:** Open a new tab.
    -   **:e filename:** Open a file for editing.
    -   **:bn, :bp:** Move to the next or previous buffer.
    -   **Ctrl + ww:** Switch between windows.
    -   **:vsp filename:** Open a file in a vertical split.

**Comment by davea42** [comment]
Additional suggestion, related to getting  a paragraph reformatted, something I use a lot
for text that will stay as plain text: (60 is just an example here):

Formatting:

`!}fmt -60`

If a paragraph starts at column n and you do this at the first letter of the paragraph the reformatted
line puts all the paragraph lines at column n.

**Comment by std-enigma** [comment]
I'm telling you this as a secret many people don't know this one:
hit two single-quote(`''`) - this will bring your cursor back to the position you jumped from
Another quality of life keybinding is: `go` which is a default binding that is equivalent to `gg0`
